File: examples/rmi/Printer/cpp/src/PrinterManagerImpl.cpp
/*
 * PrinterManagerImpl.cpp
 */

#include "PrinterManagerImpl.h"

#include <sstream>
#include <iostream>
#include <cmath>

PrinterManagerImpl::PrinterManagerImpl(CRuntime_ref runtime, std::string name) :
  runtime(runtime), s_name(name)
{
  PrinterInfos i1 = { "HPLaserJetColor", P_IDLE, "R324", true };
  printersInfo.push_back(i1);
  printersServicesImpl.push_back(PrinterImpl_ref(new PrinterImpl(printersInfo[0].name.in(),
      printersInfo[0].location.in(), printersInfo[0].color!=0 )));

  PrinterInfos i2 = { "BrotherHL-5280", P_IDLE, "R115", true };
  printersInfo.push_back(i2);
  printersServicesImpl.push_back(PrinterImpl_ref(new PrinterImpl(printersInfo[1].name.in(),
      printersInfo[1].location.in(), printersInfo[1].color!=0 )));

  PrinterInfos i3 = { "EPSON-EPL-7500", P_IDLE, "R245", false };
  printersInfo.push_back(i3);
  printersServicesImpl.push_back(PrinterImpl_ref(new PrinterImpl(printersInfo[2].name.in(),
      printersInfo[2].location.in(), printersInfo[2].color!=0 )));

  for (unsigned int i = 0; i < printersInfo.size(); i++)
    {
      //registering the Printers services
      bool ret =
          DDS_Service::register_interface<PrinterInterface, PrinterImpl>(
              printersServicesImpl[i], std::string(printersInfo[i].name.in()), 1);
      if (!ret)
        {
          std::cout << "Failed to register Printer service: "
              << printersInfo[i].name << std::endl;
          return;
        }

    }

}

PrinterManagerImpl::~PrinterManagerImpl()
{
}

void
PrinterManagerImpl::suspend()
{
  std::cout << "PrinterManager service stopped." << std::endl;
}

void
PrinterManagerImpl::resume()
{
  std::cout << "PrinterManager service activated." << std::endl;
}

DDS::String
PrinterManagerImpl::getClosestPrinter(DDS::String room)
{
  int nearestRoomIdx = 0;
  std::stringstream room_str(room);
  room_str.ignore(1);
  int room_int;
  room_str >> room_int;
  int min = 1000; //just a big value

  //finding the nearest room
  for (unsigned int i = 0; i < printersInfo.size(); i++)
    {
      std::stringstream room_str(std::string(printersInfo[i].location.in()).substr(1));
      int r;
      room_str >> r;

      int m = abs(room_int - r);
      if (m <= min)
        {
          min = m;
          nearestRoomIdx = i;
        }
    }
  return DDS::string_dup(printersInfo[nearestRoomIdx].name);
}

void
PrinterManagerImpl::getAllPrintersNames(::printer::NamesList& names)
{
  names.length((DDS::ULong)printersInfo.size());
  for (unsigned int i = 0; i < printersInfo.size(); i++)
    {
      names[i] = DDS::string_dup(printersInfo[i].name);
    }

}

void
PrinterManagerImpl::stop()
{
  std::cout
      << "The PrinterManager service will stop all the printers services ..."
      << std::endl;
  std::cout << "The PrinterManager service will stop ..." << std::endl;
  runtime->shutdown(false);
}

File: examples/rmi/Printer/java/src/dds/rmi/example/printer/PrinterManagerImpl.java
package dds.rmi.example.printer;

import java.util.Arrays;

import org.opensplice.DDS_RMI.CInterfaceManager;
import org.opensplice.DDS_RMI.CRuntime;
import org.opensplice.DDS_RMI.DDS_Service;

import printer.NamesListHolder;
import printer.PrinterInfos;
import printer.Status;
import DDS_RMI.printer.PrinterInterface;
import DDS_RMI.printer.PrinterManagerInterface;

public class PrinterManagerImpl extends PrinterManagerInterface{

    PrinterInfos[] printersInfo;
    PrinterImpl[] printersServicesImpl;
    CRuntime runtime;
    String serviceName;

    public PrinterManagerImpl(CRuntime runtime, String name){
        printersInfo = new PrinterInfos[3];
        printersServicesImpl = new PrinterImpl[3];
        printersInfo[0] = new PrinterInfos("HPLaserJetColor", Status.P_IDLE, "R324", true);
        printersServicesImpl[0] = new PrinterImpl(printersInfo[0].name, printersInfo[0].location, printersInfo[0].color);
        printersInfo[1] = new PrinterInfos("BrotherHL-5280", Status.P_IDLE, "R115", true);
        printersServicesImpl[1] = new PrinterImpl(printersInfo[1].name, printersInfo[1].location, printersInfo[1].color);
        printersInfo[2] = new PrinterInfos("EPSON-EPL-7500", Status.P_IDLE, "R245", false);
        printersServicesImpl[2] = new PrinterImpl(printersInfo[2].name, printersInfo[2].location, printersInfo[2].color);
        //registering the different Printers services
        for(int i=0; i<printersInfo.length; i++){
            boolean ret = DDS_Service.register_interface(printersServicesImpl[i], printersInfo[i].name, 1, PrinterInterface.class);
            if(!ret) {
                System.out.println("Failed to register Printer service: " + printersInfo[i].name);
                return;
            }
        }

        this.runtime = runtime;
        this.serviceName = name;

    }

    @Override
    public void suspend() {
        System.out.println("PrinterManager service stopped.");
    }

    @Override
    public void resume() {
        System.out.println("PrinterManager service activated.");

    }

    @Override
    public String getClosestPrinter(String room) {
        int nearestRoomIdx = 0;
        int roomR = Integer.valueOf(room.substring(1));
        int min = Integer.MAX_VALUE;
        //finding the nearest room
        for(int i=0; i<printersInfo.length; i++){
            int r = Integer.valueOf(printersInfo[i].location.substring(1));
            int m = Math.abs(roomR - r);
            if(m <= min){
                min = m;
                nearestRoomIdx = i;
            }
        }
        return printersInfo[nearestRoomIdx].name;
    }

    @Override
    public void getAllPrintersNames(NamesListHolder names) {
        String[] printers = new String[printersInfo.length];
        for(int i=0; i<printersInfo.length; i++){
            printers[i] = printersInfo[i].name;
        }
       names.value = Arrays.copyOf(printers, printers.length);
    }

    public void stop(){
        System.out.println("The PrinterManager service will stop all the printers services ...");
        System.out.println("The PrinterManager service will stop ...");
        runtime.shutdown(false);
    }

}
